### PR TITLE
Update Iris sample to target .NET Core and support Linux

### PR DESCRIPTION
**Pull request by gregg-miskelly**
This PR makes changes to the Iris EE sample so that the compiler will produce code that runs on .NET Core, and so that the EE can be run on either .NET Framework or .NET Core.

A big thank you to @ivanpovazan from Raincode who did the initial work for this.

Changes:
1. The projects were converted to .NET Core style "sdk" projects.
2. The work of packaging the EE into a .vsix was split out into a separate project. This is because .NET Core SDK net standard library projects don't have debug options to control if the .NET Framework or .NET Core debug engine is used. So I left the vsix project in a separate classic project instead of the SDK style.
3. Removed assumption that Console APIs are in CoreLib
4. Updated code for spawning ilasm to work with .NET Core:
    * Don't assume that it is always next to CoreLib
    * Use '-' instead of '/' with switch names as '/' doesn't work on Linux
    * Redirect standard error/out to make sure it doesn't get written back to vsdbg's std error/out on Linux
    * Pay attention to the exit code
5. Fixed Iris/Programs to run the programs on .NET Core.
6. Add a fake project to Iris\Programs to provide a project that can easily be opened in VS to make it easy to debug and allow primitive build integration. To do this I also fixed the exit code propagation in build.cmd for compile errors.
7. Add new xplat-packaging project to take care of packaging up files into a directory for supporting xplat scenarios.
8. Converted FrontEndTest to nunit

### CODE REVIEW

**Comment by gregg-miskelly** [comment]
@ivanpovazon if you have a chance, please take a look as well.

File: Iris/Programs/TicTacToe.csproj [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <StartAction>Program</StartAction>
+    <StartProgram Condition="'$(ProgramW6432)'!=''">$(ProgramW6432)\dotnet\dotnet.exe</StartProgram>
+    <StartProgram Condition="'$(ProgramW6432)'==''">$(ProgramFiles)\dotnet\dotnet.exe</StartProgram>
+    <StartArguments>$(MSBuildThisFileDirectory)TicTacToe.dll</StartArguments>
+  </PropertyGroup>
+  
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />
```
Comment (WardenGnaw) [inline Iris/Programs/TicTacToe.csproj:15]: I thought line 1 `SDK="Microsoft.NET.Sdk"` does this.

**Review by WardenGnaw — APPROVED** [review]

**Comment by ivanpovazan** [comment]
> @ivanpovazon if you have a chance, please take a look as well.
will do tomorrow during the day :)

File: Iris/Programs/TicTacToe.csproj (hunk 1 above)
Comment (gregg-miskelly) [reply Iris/Programs/TicTacToe.csproj:15]: The problem is order -- if you don't have this Import statement, then these targets will get overwritten by the default target. So instead you need to include the default first, and then overwrite them.

Comment (gregg-miskelly) [reply Iris/Programs/TicTacToe.csproj:15]: Its the same reason for it [here](https://github.com/microsoft/MIEngine/blob/4221ef54433d297aa88f49c027c7d8f9bc35f95e/src/MIDebugPackage/MIDebugPackage.csproj#L274) BTW.

File: Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs [hunk 2]
```diff
@@ -46,11 +46,7 @@ internal class DebugCompilerContext : CompilerContext
             List<DkmClrLocalVariableInfo> generatedLocals,
             string assignmentLValue,
             bool argumentsOnly)
-#if NETCOREAPP
-            : this(scope.Session.Importer, reader, CompilationFlags.NoDebug | CompilationFlags.WriteDll | CompilationFlags.NetCore)
```
Comment (ivanpovazan) [inline Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs]: A minor detail... If `CompilationFlags.NetCore` is not set on CompilerContext wouldn't it limit the PEemitter to always look for netframework ilasm on Windows (see [here](https://github.com/microsoft/ConcordExtensibilitySamples/blob/9641ba83e05b0d1f6b7f8c6c645a3c2b0beb9eb4/Iris/IrisCompiler/BackEnd/PeEmitter.cs#L78)) breaking the VSCode-only scenario?

Please correct me, if I am missing something

**Comment by gregg-miskelly** [comment]
> > @ivanpovazon if you have a chance, please take a look as well.
> 
> @gregg-miskelly
> there is a minor typo in my name but no hard feelings :) I will check out all the changes tomorrow during the day
> Thanks!

Oops. Sorry about that! I fixed it in the PR description.

File: Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs (hunk 2 above)
Comment (gregg-miskelly) [reply Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs]: I left `CompilationFlags.NetCore` in as it seemed like a useful concept, but it actually doesn't do much -- [this line](https://github.com/microsoft/ConcordExtensibilitySamples/blob/9641ba83e05b0d1f6b7f8c6c645a3c2b0beb9eb4/Iris/IrisCompiler/BackEnd/PeEmitter.cs#L85) will check if the framework-adjacent ilasm.exe exists, and if not it will go down the 'running on .NET Core' case regardless of the flag.

File: Iris/FrontEndTest/TestCompilerContext.cs [hunk 3]
```diff
@@ -50,6 +50,11 @@ internal sealed class TestCompilerContext : CompilerContext
 
         public static TestCompilerContext Create(string compiland, GlobalSymbolList globals, CompilationFlags flags)
         {
+#if NETCOREAPP
+            flags |= CompilationFlags.NetCore;
+            flags |= CompilationFlags.WriteDll;
```
Comment (plnelson) [inline Iris/FrontEndTest/TestCompilerContext.cs:55]: >flags |= CompilationFlags.WriteDll; [](start = 12, length = 35)

Why is this needed for TestCompilerContext?  We should be writing text instead of a PE.

File: Iris/ic/CompilerRunner.cs [hunk 4]
```diff
@@ -101,6 +105,12 @@ private static bool TryParseArgs(string[] args, out string sourceFile, out Compi
             if (sourceFile == null)
                 return false;
 
+
+#if NETCOREAPP
```
Comment (plnelson) [inline Iris/ic/CompilerRunner.cs:109]: If I'm understanding the .csproj correctly, aren't we always building "ic" with .NET 5 now?  If so why do we have conditional compilation for .NET Core?

File: Iris/IrisCompiler/CompilerContext.cs [hunk 5]
```diff
@@ -112,12 +111,18 @@ public void AddIntrinsics()
...
                 ImportMethod(fp, "str", mscorlib, "System.Int32", "ToString", true, tString, noParams);
                 ImportMethod(fp, "strcmp", mscorlib, "System.String", "Compare", false, tInt, new IrisType[] { tString, tString });
-                ImportMethod(fp, "writeln", mscorlib, "System.Console", "WriteLine", false, tVoid, new IrisType[] { tString });
+
+                ImportedModule consoleLib = ReferenceConsoleLib();
+                if (consoleLib != null)
+                {
+                    ImportMethod(fp, "readln", consoleLib, "System.Console", "ReadLine", false, tString, noParams);
+                    ImportMethod(fp, "writeln", consoleLib, "System.Console", "WriteLine", false, tVoid, new IrisType[] { tString });
+                }
```
Comment (plnelson) [inline Iris/IrisCompiler/CompilerContext.cs:122]: >} [](start = 16, length = 1)

`else` Import `readln` and `writeln` from mscorlib.

File: Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs (hunk 2 above)
Comment (ivanpovazan) [reply Iris/IrisExtension/ExpressionCompiler/DebugCompilerContext.cs]: Ah I see, thanks for clarification

File: Iris/IrisCompiler/BackEnd/PeEmitter.cs [hunk 6]
```diff
@@ -63,30 +64,66 @@ public void Flush()
...
+                if (ilasmPath == null || !File.Exists(ilasmPath))
+                {
+                    string thisDir = Path.GetDirectoryName(GetType().Assembly.Location);
+                    string ilasmFileName = isWindows ? "ilasm.exe" : "ilasm";
+                    ilasmPath = Path.Combine(thisDir, ilasmFileName);
+
+                    if (!File.Exists(ilasmPath))
+                        throw new FileNotFoundException("ilasm cannot be found, make sure netcore ilasm is present in the same directory as IrisCompiler.dll");
+
+                    if (!string.IsNullOrEmpty(debug)) // netcore ilasm requires explicitly specifying pdb format
```
Comment (plnelson) [inline Iris/IrisCompiler/BackEnd/PeEmitter.cs:94]: >) [](start = 52, length = 1)

Should this specify full PDBs if building for Desktop CLR?  AFAIK Roslyn creates full PDBs unless building for .NET Core.

File: Iris/IrisCompiler/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -8,12 +8,7 @@
 // General Information about an assembly is controlled through the following 
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
-[assembly: AssemblyTitle("IrisParser")]
```
Comment (plnelson) [inline Iris/IrisCompiler/Properties/AssemblyInfo.cs:11]: Wow, this file had garbage.  Thanks for the cleanup!

File: Iris/IrisCompiler/BackEnd/PeEmitter.cs (hunk 6 above)
Comment (ivanpovazan) [reply Iris/IrisCompiler/BackEnd/PeEmitter.cs:94]: Prior to .NET 5.0 ilasm for .NET Core did not have support for producing PDB at all. 
With the 5.0 version it received the support for Portable PDB generation.
I've introduced the `-pdbfmt` flag to emphasize this but also to leave it open in case Microsoft decides to support other variants/formats (embedded, full, legacy-classic PDBs)

File: Iris/ic/CompilerRunner.cs (hunk 4 above)
Comment (gregg-miskelly) [reply Iris/ic/CompilerRunner.cs:109]: I left that in in case anyone wanted to modify the project to target desktop. But you are correct that it will always be used. I think I will leave a comment.

File: Iris/IrisCompiler/CompilerContext.cs (hunk 5 above)
Comment (gregg-miskelly) [reply Iris/IrisCompiler/CompilerContext.cs:122]: No. If somehow that method returns false then it will not be in mscorlib.

File: Iris/IrisCompiler/BackEnd/PeEmitter.cs (hunk 6 above)
Comment (gregg-miskelly) [reply Iris/IrisCompiler/BackEnd/PeEmitter.cs:94]: FYI Patrick your statement isn't quite correct: Roslyn (or maybe it is actually msbuild targets?) *defaults* to full PDBs on desktop, but the options are independent. So it is possible to get a portable PDB when targeting desktop or a full PDB when targeting .NET Core (assuming you are on Windows).

Regardless, it would be very weird to get into this code path if you are targeting desktop since we will generally find ilasm.exe in the above code path in this case. So I don't think it makes sense to change it.

File: Iris/IrisCompiler/CompilerContext.cs (hunk 5 above)
Comment (plnelson) [reply Iris/IrisCompiler/CompilerContext.cs:122]: Ah, I understand.  ReferenceConsoleLib will return mscorlib on Desktop CLR.  It wasn't clear to me from the name.

---
In reply to: [562862041](https://github.com/microsoft/ConcordExtensibilitySamples/pull/65#discussion_r562862041) [](ancestors = 562862041)

File: Iris/IrisCompiler/BackEnd/PeEmitter.cs (hunk 6 above)
Comment (plnelson) [reply Iris/IrisCompiler/BackEnd/PeEmitter.cs:94]: Right, I see the code path shouldn't be used on desktop.

---
In reply to: [562864868](https://github.com/microsoft/ConcordExtensibilitySamples/pull/65#discussion_r562864868) [](ancestors = 562864868)

File: Iris/FrontEndTest/TestCompilerContext.cs (hunk 3 above)
Comment (gregg-miskelly) [reply Iris/FrontEndTest/TestCompilerContext.cs:55]: Yup. Doesn't seem to be needed. I will remove it.

**Comment by plnelson** [comment]
👍 LGTM

**Review by plnelson — APPROVED** [review]
:shipit:
